### ISSUE TITLE
Single misbehaving client can stall poller

### ISSUE THREAD

**Issue by shsmith** [spesmilo/electrum-server#106]
The do_handshake operation is single-threaded and runs in the main poller thread. If a misbehaving client causes this process to take too long (or forever) it prevents the server from accepting new connections and also stops processing existing connections. 

I think we need to call settimeout on the connection prior to calling do_handshake, like this:
```
                timeout = session._connection.gettimeout()
                session._connection.settimeout(60)
                session._connection.do_handshake()
                session._connection.settimeout(timeout)
~~~
It would be even better if there was a non-blocking way to do the handshake, or possibly the handshake can run in a worker thread that does not block the main poller thread.

**Comment by ecdsa** [comment]
can you write a pull request?

**Comment by shsmith** [comment]
See PR #107

**Comment by ecdsa** [comment]
thanks!

**Comment by ecdsa** [comment]
see my comment on the patch: setting a timeout actually makes the socket blocking.
can you explain in which conditions a client can stall the poller? is this reproducible?

**Comment by ecdsa** [comment]
PR reverted in c12cc1ce62471a145950da6010e2a2933b1c3651

**Comment by shsmith** [comment]
Thank you for your help with this.
I definitely had poller stalls before my original PR but maybe there was a different root cause than do_handshake.
I pulled your latest and will keep an eye out for any more poller stalls. 
If I do witness a new poller stall, is there some way (maybe via run_electrum_server debug) that I can determine what operation is blocking the thread?

**Comment by shsmith** [comment]
I was wrong about do_handshake blocking the poller.  
I think you fixed it in 6052ee3667d35e96c46b2312207998d2274e924a. 
My apologies.
